From 487b8bb6d145f07c004084af90aed78279b1496b Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Tue, 23 Aug 2022 12:04:26 +0200
Subject: [PATCH 0772/1010] evl/thread: duly mark SIGDEBUG cause on direct
 signal

The sigval information was not bearing the SIGDEBUG marker in case we
signaled from in-band context directly. Set the marker in the common
code path shared between oob and in-band contexts to fix this.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/thread.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/kernel/evl/thread.c b/kernel/evl/thread.c
index 606f24b902..4dc3aca39b 100644
--- a/kernel/evl/thread.c
+++ b/kernel/evl/thread.c
@@ -1304,10 +1304,11 @@ static void do_inband_signal(struct evl_thread *thread, int signo, int sigval)
 		memset(&si, '\0', sizeof(si));
 		si.si_signo = signo;
 		si.si_code = SI_QUEUE;
-		si.si_int = sigval;
+		si.si_int = sigval | sigdebug_marker;
 		send_sig_info(signo, &si, p);
-	} else
+	} else {
 		send_sig(signo, p, 1);
+	}
 }
 
 static void sig_irqwork(struct irq_work *work)
@@ -1336,7 +1337,7 @@ void evl_signal_thread(struct evl_thread *thread, int sig, int arg)
 	init_irq_work(&sigd->work, sig_irqwork);
 	sigd->thread = thread;
 	sigd->signo = sig;
-	sigd->sigval = sig == SIGDEBUG ? arg | sigdebug_marker : arg;
+	sigd->sigval = arg;
 
 	evl_get_element(&thread->element);
 	/* Cannot fail, irq_work is local to this call. */
-- 
2.30.2

